File: model/layers/output_head.py
"""
model/layers/output_head.py

2-layer MLP: flattened sync matrix → ReLU → linear → softmax → class probabilities.

Produces the final classification distribution after the tick loop converges.
"""

from __future__ import annotations

import torch
import torch.nn as nn
import torch.nn.functional as F


class OutputHead(nn.Module):
    """Maps a flattened sync matrix to a probability distribution over classes.

    Parameters
    ----------
    n_neurons : int
        Number of neurons N. Input dim = N².
    hidden_dim : int
        Width of the hidden layer.
    n_classes : int
        Number of output classes.
    """

    def __init__(self, n_neurons: int, hidden_dim: int, n_classes: int) -> None:
        super().__init__()
        sync_flat = n_neurons * n_neurons
        self.fc1 = nn.Linear(sync_flat, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, n_classes)

    def forward(self, sync_matrix: torch.Tensor) -> torch.Tensor:
        """Compute class probabilities from sync matrix.

        Parameters
        ----------
        sync_matrix : torch.Tensor
            Shape (N, N).

        Returns
        -------
        torch.Tensor
            Shape (n_classes,) — probability distribution summing to 1.
        """
        x = F.relu(self.fc1(sync_matrix.flatten()))
        return F.softmax(self.fc2(x), dim=-1)

    def logits(self, sync_matrix: torch.Tensor) -> torch.Tensor:
        """Return raw logits (pre-softmax) for use with CrossEntropyLoss."""
        x = F.relu(self.fc1(sync_matrix.flatten()))
        return self.fc2(x)

File: src/lib/model/output-head.ts
import { matVec, softmax, flatten } from "./linalg";

/**
 * OutputHead
 *
 * MLP that maps the sync matrix (N×N pairwise correlations) to a probability
 * distribution over a fixed vocabulary of classes.
 *
 * Architecture:
 *   h = ReLU(W1 · flatten(syncMatrix) + b1)   shape: (hiddenDim,)
 *   logits = W2 · h + b2                       shape: (nClasses,)
 *   probs  = softmax(logits)                   shape: (nClasses,)
 *
 * Weights are deterministically initialised with seeded Xavier uniform,
 * biases at zero.
 */
export interface OutputHeadConfig {
  /** Number of neurons N; input is N×N = N² features */
  nNeurons: number;
  /** Number of hidden units */
  hiddenDim: number;
  /** Number of output classes */
  nClasses: number;
}

export class OutputHead {
  readonly config: OutputHeadConfig;

  /** W1: (hiddenDim × N²) */
  readonly W1: number[][];
  /** b1: (hiddenDim,) — initialised to zero */
  readonly b1: number[];
  /** W2: (nClasses × hiddenDim) */
  readonly W2: number[][];
  /** b2: (nClasses,) — initialised to zero */
  readonly b2: number[];

  constructor(config: OutputHeadConfig, seed = 42) {
    this.config = config;
    const inputDim = config.nNeurons * config.nNeurons;
    this.W1 = initWeights(config.hiddenDim, inputDim, inputDim, seed);
    this.b1 = new Array(config.hiddenDim).fill(0);
    this.W2 = initWeights(config.nClasses, config.hiddenDim, config.hiddenDim, seed + 1);
    this.b2 = new Array(config.nClasses).fill(0);
  }

  /**
   * Forward pass.
   * @param syncMatrix  N×N pairwise correlation matrix (entries in [-1, 1])
   * @returns           Probability distribution over nClasses (sums to 1.0)
   */
  forward(syncMatrix: number[][]): number[] {
    const x = flatten(syncMatrix);

    // Hidden layer: ReLU(W1·x + b1)
    const hidden = matVec(this.W1, x)
      .map((v, i) => Math.max(0, v + this.b1[i]));

    // Output logits: W2·h + b2
    const logits = matVec(this.W2, hidden)
      .map((v, i) => v + this.b2[i]);

    return softmax(logits);
  }
}

// ── Weight initialisation ─────────────────────────────────────────────────────

function initWeights(rows: number, cols: number, fanIn: number, seed: number): number[][] {
  const limit = Math.sqrt(6 / fanIn);
  let state = seed >>> 0;
  function next(): number {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
    return state / 0xffffffff;
  }
  return Array.from({ length: rows }, () =>
    Array.from({ length: cols }, () => (next() * 2 - 1) * limit),
  );
}
